Add/34000 attribute select control

louwie17 opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This PR enables the select2 typeahead field for the attribute dropdown and the terms dropdown within the individual product attributes.
Currently I added a threshold for the attribute dropdown of 20, if more it will show the async typeahead field. This is similar to what is done for the category dropdown in the product list page. I could remove this condition if we don't think this is necessary.

https://user-images.githubusercontent.com/2240960/191288330-704e81fa-7221-46a2-9fe8-1feb046b4e9e.mp4

Partially address #34000

### How to test the changes in this Pull Request:

1. Load and build this PR and finish the onboarding, adding several products and attributes (an easy way is to import the [sample products CSV](https://github.com/woocommerce/woocommerce/tree/trunk/plugins/woocommerce/sample-data)).
2. Either edit an existing product or create a new one and select the **Attributes** tab
3. Assume you still have attributes less then 20, the normal dropdown should show.
4. Add the color attribute and add some terms. Notice how the terms are loaded on the fly and filtered as you type.
5. Now either add more then 20 different attributes (would probably have to be done manually) or add this filter to your `mu-plugin` -> `add_filter( 'woocommerce_attribute_taxonomy_filter_threshold', function() { return 1; } )`
6. It should now show a **Add custom attribute** button and a typeahead dropdown in the Attributes tab.
7. Test out the **Add custom attribute** button, it should add the custom item.
8. Now search for existing attributes and upon selection it should add the attribute to the product
9. Try searching for your previously selected item, it should be disabled now.
10. Now remove the attribute, the dropdown should not show it as disabled anymore.
11. Saving the attributes and persisting by refreshing the page should work the same as usual.

### Other information:

-   [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [ ] Have you successfully run tests with your changes locally?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm changelog add --filter=<project>`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

louwie17 committed:
Replace preloaded attribute and term dropdowns with async typeahead dropdowns

louwie17 committed:
Revert accidenatal change.

louwie17 committed:
Add condition to show the old dropdown in certain situations

louwie17 committed:
Add changelog

louwie17 commented on plugins/woocommerce/client/legacy/js/admin/meta-boxes-product.js:
```diff
@@ -474,6 +475,37 @@ jQuery( function ( $ ) {
 		}
+	}
+
+	$('select.wc-attribute-search').on('select2:select', function (e) {
```
callback for when user selects something from the dropdown field.

louwie17 commented on plugins/woocommerce/client/legacy/js/admin/meta-boxes-product.js:
```diff
@@ -474,6 +475,37 @@ jQuery( function ( $ ) {
+	});
+
+	// Add rows.
+	$( 'button.add_attribute' ).on( 'click', function () {
```
Same callback as before just moved down and using the `add_attribute` function

louwie17 commented on plugins/woocommerce/includes/admin/class-wc-admin-assets.php:
```diff
@@ -135,21 +135,23 @@ public function admin_scripts() {
+					'search_customers_nonce'          => wp_create_nonce( 'search-customers' ),
+					'search_categories_nonce'         => wp_create_nonce( 'search-categories' ),
+					'search_taxonomy_terms_nonce'     => wp_create_nonce( 'search-taxonomy-terms' ),
+					'search_product_attributes_nonce' => wp_create_nonce( 'search-product-attributes' ),
```
I only added the above two, but had to re-indent everything to re-align correctly.

louwie17 commented on plugins/woocommerce/includes/class-wc-ajax.php:
```diff
@@ -1727,6 +1729,95 @@ public static function json_search_categories() {
+		$found_product_categories = array();
+
+		foreach ( $attributes as $attribute_obj ) {
+			if ( ! $search_text || false !== stripos( $attribute_obj->attribute_label, $search_text ) ) {
```
I end up resorting to filtering by getting the entire list (which is cached) and just sorting through this on the back-end.

I browsed around for a while and it turns out we do not have any logic in place to search for attribute taxonomies. We seem to always retrieve the entire list and cache this. To keep things simple I did it this way.
But I would foresee us needing this in the future and also adding this functionality to the Woo REST api (as it currently doesn't support this either).

I could still add this as part of this issue, but it felled slightly out of scope, as it will require the addition of a whole new attribute taxonomy Query class.

octaedro commented on plugins/woocommerce/client/legacy/js/admin/wc-enhanced-select.js:
```diff
@@ -304,6 +304,98 @@ jQuery( function( $ ) {
+				// Ajax category search boxes
+				$( ':input.wc-taxonomy-term-search' ).filter( ':not(.enhanced)' ).each( function() {
+					var return_format = $( this ).data( 'return_id' ) ? 'id' : 'slug';
+					var minimumInputLength = 3;
+
+					if ( $( this ).data( 'minimum_input_length' ) !== undefined ) {
+						minimumInputLength = $( this ).data( 'minimum_input_length' );
+					}
```
Can we use here a ternary operator as the line before?
It would look like this:
```
var minimumInputLength = $( this ).data( 'minimum_input_length' ) !== undefined ? $( this ).data( 'minimum_input_length' ) : 3;
```
Another good option would be:
```
var minimumInputLength = $( this ).data( 'minimum_input_length' ) ?? 3;
```

louwie17 replied:
Fixed as part of: https://github.com/woocommerce/woocommerce/pull/34744/commits/9cf74f8da34fd29d5ac415bf6901edc952bb3547

octaedro reviewed (commented):
You did a really nice job here @louwie17! This is testing well on my end and the code looks mostly good. I just left a small comment.
As a side note, we should probably do something similar with the attribute terms since it seems like we show a maximum of 6.

![Screen Capture on 2022-09-29 at 16-24-41](https://user-images.githubusercontent.com/1314156/193124886-9c5c8159-c8bb-4b11-a9b5-6fd4c7905aac.gif)

louwie17 committed:
Use ternary condition

louwie17 committed:
Move mininumInputLength down to follow the same format as the other filters

louwie17 commented:
@octaedro thanks for the review, I addressed your feedback.

> As a side note, we should probably do something similar with the attribute terms since it seems like we show a maximum of 6.

I am kinda confused with what you mean here? Where are we showing a maximum of 6? the page size is 50.

louwie17 committed:
Fix the minimum input length

louwie17 committed:
Update the select all to work with the async terms field

louwie17 committed:
Make sure empty attributes are shown and suppress query filter

octaedro reviewed (dismissed):
Thank you @louwie17 for the changes. LGTM 🚀

louwie17 committed:
Fix code sniff error

octaedro reviewed (approved):
🚀

florianmastacan commented:
> @octaedro thanks for the review, I addressed your feedback.
> 
> > As a side note, we should probably do something similar with the attribute terms since it seems like we show a maximum of 6.
> 
> I am kinda confused with what you mean here? Where are we showing a maximum of 6? the page size is 50.

Hi there, may I ask why the page size was limited to 50? We have a client who is using 186 colours containing the "green" word and thus when he tries to add attributes to a product in order to build the relevant variations he cannot get to the items from positions 51+ and above...
